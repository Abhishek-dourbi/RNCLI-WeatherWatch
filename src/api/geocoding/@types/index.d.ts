export interface GeocodingResponse {
    id: number;
    name: string;
    latitude: number;
    longitude: number;
}