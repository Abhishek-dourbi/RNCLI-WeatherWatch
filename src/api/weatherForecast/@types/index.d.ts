import { WeatherCode } from "../../../helpers/getWeatherImage";

export interface WeatherApiParams {
    latitude: number;
    longitude: number;
}

interface CurrentDayWeather {
    time: string;
    weather_code: WeatherCode;
    temperature_2m: number;
}

interface DailyWeather {
    time: string[],
    weather_code: WeatherCode[];
    temperature_2m_max: number[];
    temperature_2m_min: number[];
}

interface WeatherApiResponse {
    latitude: number;
    longitude: number;
    current: CurrentDayWeather;
    daily: DailyWeather;
}