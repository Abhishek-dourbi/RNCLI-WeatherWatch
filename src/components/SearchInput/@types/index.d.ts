export interface SearchInputProps {
    onChangeText(query: string): void;
    onFocus(): void;
}