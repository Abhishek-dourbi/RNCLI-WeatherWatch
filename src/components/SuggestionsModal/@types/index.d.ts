import React, { ReactNode } from "react";

export interface SuggestionModalProps<T> {
    data: T[];
    renderText(item: T, index: number): React.JSX.Element;
    isLoading: boolean;
}

interface SuggestionModalRef {
    show(): void;
    hide(): void;
    close(): void;
}