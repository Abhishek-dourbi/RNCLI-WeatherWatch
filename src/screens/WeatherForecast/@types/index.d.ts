import { WeatherCode } from "../../../helpers/getWeatherImage";

interface DailyWeatherStateData {
    date: string,
    weatherCode: WeatherCode | null,
    maxTemp: number,
    minTemp: number,
    averageTemp: string
}

export interface WeatherForecastState {
    currentTime: string;
    currentTemp: number;
    currentWeatherCode: WeatherCode | null;
    
    dailyWeatherData: DailyWeatherStateData[]
}

interface GeoCodingState {
    latitude: number;
    longitude: number;
    locationName: string;
}