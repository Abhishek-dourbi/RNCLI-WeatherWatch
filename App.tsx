import React from 'react';
import {StyleSheet, Text, View} from 'react-native';
import WeatherForecast from './src/screens/WeatherForecast';

export default function App() {
  return (
    <WeatherForecast />
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    alignItems: 'center',
    justifyContent: 'center',
  },
});
