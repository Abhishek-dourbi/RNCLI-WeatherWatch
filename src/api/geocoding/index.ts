import axios from "axios";
import { GeocodingResponse } from "./@types";

export const getLocation = async (query: string): Promise<GeocodingResponse[]> => {
    try {
        const res = await axios.get(
            'https://geocoding-api.open-meteo.com/v1/search',
            {
                params: {
                    name: query
                }
            }
        );

        return res?.data?.results || [];
    } catch(err: any) {
        throw new Error(err.message || 'Something went wrong');
    }
  };
  