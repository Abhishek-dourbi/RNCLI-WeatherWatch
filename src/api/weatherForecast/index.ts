import axios from "axios";
import { WeatherApiParams, WeatherApiResponse } from "./@types";

export const getWeatherData = async (payload: WeatherApiParams): Promise<WeatherApiResponse> => {
    try {
        const res = await axios.get(
            'https://api.open-meteo.com/v1/forecast',
            {
                params: {
                    current: 'temperature_2m,weather_code',
                    daily: 'weather_code,temperature_2m_max,temperature_2m_min',
                    timezone: 'auto',
                    ...payload
                }
            }
        );

        return res?.data;
    } catch(err: any) {
        throw new Error(err.message || 'Something went wrong');
    }
  };
  