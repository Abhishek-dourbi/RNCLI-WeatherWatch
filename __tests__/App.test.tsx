import React from 'react';
import {
  render,
  waitFor,
  fireEvent,
} from '@testing-library/react-native';
import App from '../App';
import {getWeatherData} from '../src/api/weatherForecast';
import {getLocation} from '../src/api/geocoding';
import {WeatherApiResponse} from '../src/api/weatherForecast/@types';
import {GeocodingResponse} from '../src/api/geocoding/@types';

jest.mock('../src/api/weatherForecast', () => ({
  getWeatherData: jest.fn(),
}));

jest.mock('../src/api/geocoding', () => ({
  getLocation: jest.fn(),
}));

jest.mock('../src/helpers/getWeatherImage', () =>
  jest.fn(() => ({
    image: 'https://openweathermap.org/img/wn/03d@2x.png',
    description: 'Cloudy',
  })),
);

const mockGetWeatherData = getWeatherData as jest.MockedFunction<
  typeof getWeatherData
>;
const mockGetLocation = getLocation as jest.MockedFunction<typeof getLocation>;

describe('App Component with WeatherForecast', () => {
  const mockWeatherData = {
    latitude: 28.6139,
    longitude: 77.209,
    current: {
      time: '2024-09-10T12:00:00Z',
      temperature_2m: 18,
      weather_code: '0',
    },
    daily: {
      time: [
        '2024-09-10',
        '2024-09-11',
        '2024-09-12',
        '2024-09-13',
        '2024-09-14',
        '2024-09-15',
        '2024-09-16',
      ],
      weather_code: ['0', '1', '2', '3', '45', '48', '51'],
      temperature_2m_max: [30, 31, 29, 32, 35, 31, 29],
      temperature_2m_min: [22, 21, 23, 24, 27, 20, 19],
    },
  };

  const mockLocationData = [
    {id: 1, name: 'New Delhi', latitude: 28.6139, longitude: 77.209},
  ];

  beforeEach(() => {
    mockGetWeatherData.mockResolvedValue(mockWeatherData as WeatherApiResponse);
    mockGetLocation.mockResolvedValue(mockLocationData as GeocodingResponse[]);
  });

  it('should render WeatherForecast and display the weather data', async () => {
    const {getByText, getByTestId} = render(<App />);

    await waitFor(() => {
      expect(getByText('New Delhi')).toBeTruthy();
      expect(getByText('18Â°')).toBeTruthy();
      expect(getByText('Cloudy')).toBeTruthy();
      expect(getByTestId('current-forecast-image')).toBeTruthy();
    });
  });

  it('should handle location input and display suggestions', async () => {
    const {getByPlaceholderText, getByText} = render(<App />);

    const input = getByPlaceholderText('Search...');

    fireEvent.changeText(input, 'Del');

    await waitFor(() => {
      expect(getByText('New Delhi')).toBeTruthy();
    });
  });

  it('should match the snapshot', () => {
    const {toJSON} = render(<App />);
    expect(toJSON()).toMatchSnapshot();
  });
});
