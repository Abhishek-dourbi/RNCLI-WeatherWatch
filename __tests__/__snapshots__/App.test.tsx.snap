// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component with WeatherForecast should match the snapshot 1`] = `
<RCTSafeAreaView
  style={
    {
      "backgroundColor": "#20A7DB",
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
    testID="loading-indicator"
  >
    <ActivityIndicator
      color="#fff"
      size="large"
    />
  </View>
</RCTSafeAreaView>
`;
